File: cmd/serf-tui/tool_renderers.go
package main

import (
	"encoding/json"
	"strconv"
	"strings"
	"time"
)

// ToolArgs is a decoded JSON args map with a convenience Str() accessor.
type ToolArgs map[string]any

// toolArgsFromJSON decodes a JSON object string into ToolArgs.
// Returns an empty map on any error or empty input.
func toolArgsFromJSON(s string) ToolArgs {
	if s == "" {
		return ToolArgs{}
	}
	var args ToolArgs
	if err := json.Unmarshal([]byte(s), &args); err != nil {
		return ToolArgs{}
	}
	return args
}

// Str returns the string value for key, or "" if absent or not a string.
func (a ToolArgs) Str(key string) string {
	if v, ok := a[key].(string); ok {
		return v
	}
	return ""
}

// ToolRenderer describes how to render a single tool call in the TUI.
type ToolRenderer struct {
	// Verb returns the short action word (e.g. "read", "edit", "shell").
	Verb func(args ToolArgs) string
	// Target returns the primary subject (e.g. file path, command).
	Target func(args ToolArgs) string
	// Result returns the compact outcome text for a completed call.
	Result func(args ToolArgs, output, errStr string, dur time.Duration) string
	// Body renders an expanded multi-line body. May be nil (no body).
	Body func(args ToolArgs, output string, w int) string
	// ExpandedByDefault causes the body to show without user interaction.
	ExpandedByDefault bool
}

// toolRenderers is the registry of per-tool renderers.
var toolRenderers = map[string]ToolRenderer{}

// lookupToolRenderer returns the renderer for tool.  Falls back to an
// MCP-style renderer (provider__operation) or an unknown-tool renderer.
// Always returns ok=true — the fallback is the last resort.
func lookupToolRenderer(tool string) (ToolRenderer, bool) {
	if r, ok := toolRenderers[tool]; ok {
		return r, true
	}
	if strings.Contains(tool, "__") {
		return mcpFallbackRenderer(tool), true
	}
	return unknownToolRenderer(tool), true
}

func mcpFallbackRenderer(tool string) ToolRenderer {
	provider, op, _ := strings.Cut(tool, "__")
	return ToolRenderer{
		Verb:   func(_ ToolArgs) string { return provider },
		Target: func(_ ToolArgs) string { return op },
		Result: func(_ ToolArgs, _ string, errStr string, _ time.Duration) string {
			if errStr != "" {
				return "error"
			}
			return "ok"
		},
	}
}

func unknownToolRenderer(tool string) ToolRenderer {
	return ToolRenderer{
		Verb:   func(_ ToolArgs) string { return tool },
		Target: func(_ ToolArgs) string { return "" },
		Result: func(_ ToolArgs, _ string, errStr string, _ time.Duration) string {
			if errStr != "" {
				return "error"
			}
			return "ok"
		},
	}
}

// formatLineCount returns "1 line" or "N lines".
func formatLineCount(n int) string {
	if n == 1 {
		return "1 line"
	}
	return strconv.Itoa(n) + " lines"
}

// shortID returns the first 8 characters of an ID string.
func shortID(id string) string {
	if len(id) <= 8 {
		return id
	}
	return id[:8]
}

// diffResultText counts +/- lines from a unified-diff output and returns a
// compact summary such as "3 +/2 -", "4 added", "2 removed", or "ok".
func diffResultText(_ ToolArgs, output, errStr string, _ time.Duration) string {
	if errStr != "" {
		return "error"
	}
	plus, minus := 0, 0
	for _, line := range strings.Split(output, "\n") {
		if strings.HasPrefix(line, "+") && !strings.HasPrefix(line, "+++") {
			plus++
		}
		if strings.HasPrefix(line, "-") && !strings.HasPrefix(line, "---") {
			minus++
		}
	}
	switch {
	case plus > 0 && minus > 0:
		return strconv.Itoa(plus) + " +/" + strconv.Itoa(minus) + " -"
	case plus > 0:
		return strconv.Itoa(plus) + " added"
	case minus > 0:
		return strconv.Itoa(minus) + " removed"
	default:
		return "ok"
	}
}

func init() {
	// read_file
	toolRenderers["read_file"] = ToolRenderer{
		Verb:   func(_ ToolArgs) string { return "read" },
		Target: func(args ToolArgs) string { return args.Str("file_path") },
		Result: func(_ ToolArgs, output, errStr string, _ time.Duration) string {
			if errStr != "" {
				return "error"
			}
			lines := strings.Count(output, "\n") + 1
			return formatLineCount(lines)
		},
	}
}

File: cmd/serf-hub/frontend/src/panes/session/transcript/toolRenderers.ts
// The tool-renderer registry: one descriptor per tool name (or a job_*-style
// predicate family), keyed off ItemModel.toolName for "commandExecution"
// items. Wave 4 T1 ships the registry + the raw-output fallback; T3
// registers the real per-tool descriptors (read/grep/ls/glob/shell/diff/
// patch/web fetch+search/delegate/job_*/ask_user/sandbox escalation).
import { type ComponentType, createElement, Fragment } from "react";
import type { ItemModel, ThreadModel } from "../../../protocol/model";
import type { ToolIconKind } from "../../../widgets";
import { MCPToolArguments } from "./MCPToolArguments";
import { RawToolOutput } from "./RawToolOutput";

export interface ToolRenderProps {
  item: ItemModel;
  live: boolean;
  // The enclosing session's ref (ToolCallItem's own sessionRef prop,
  // threaded straight through) - undefined for the same reasons ItemRenderProps'
  // own sessionRef is (Session.tsx not yet in the render path, e.g. a future
  // caller that never sets it). The subagent-transcript body is the one
  // consumer today (subagentModule.tsx's openTranscript): a delegate row's
  // "Open transcript" link needs it to record which session to return to
  // (kata 0pzz - a subagent transcript is a child of a specific parent
  // session, and the pane it opens must be able to say so and offer a way
  // back, not just leave the reader to remember it).
  sessionRef?: string;
}

export interface ToolRendererDescriptor {
  match: string | ((toolName: string) => boolean); // exact name or predicate (job_* family)
  summary(item: ItemModel): string; // one-line purpose-first summary
  // The tool-FAMILY glyph riding inline at the start of the row's tool-use
  // line (widgets/toolicon), so the kind of work - shell vs file read vs
  // edit vs web - is scannable down a run of calls without reading the
  // summary text. Optional: a descriptor without one renders no icon, and
  // every unregistered tool - including every MCP tool - inherits the
  // DEFAULT_DESCRIPTOR's generic wrench.
  icon?: ToolIconKind;
  // Fixed-width summary text. The summary face is sans by default (Jesse's
  // review call: fixed-width everywhere made every tool read like a
  // terminal); shell opts in because its summary IS a command.
  monoSummary?: boolean;
  body?: ComponentType<ToolRenderProps>; // expanded content; default raw output
  autoExpand?(item: ItemModel): boolean; // e.g. shell on nonzero exit
  // failed is a TOOL-SPECIFIC failure signal, OR'd with the generic one
  // ToolCallItem derives from ItemModel.error/status. It exists because a
  // clean tool RESULT can still report a failed action: a shell command that
  // ran and exited nonzero carries no ItemModel.error at all (the wire stamps
  // status "completed"), yet it is exactly what the reader needs the failure
  // glyph for. Mirrors the legacy renderer's own toolLooksGood, which likewise
  // treated a nonzero exit_code as not-good (renderer-format.js:593).
  failed?(item: ItemModel): boolean;
  // detail is secondary fact the row must keep REACHABLE without making it the
  // headline - rendered as the row's hover title. The shell exit code is the
  // motivating case (A2: "exit 1" stops being the failure signal, the glyph is).
  detail?(item: ItemModel): string | undefined;
  // suppress removes the whole tool-call row from the transcript when true -
  // no summary, no body, nothing (ToolCallItem renders null). Used for a
  // task_list `action:"view"` (a read that legacy renders nothing for) and a
  // malformed non-mutation call: the same "fully suppressed - no card, no
  // divider, no tool-call row" the legacy renderer applied. An errored call is
  // never suppressed here, so its error still surfaces via the generic
  // failed-row treatment in ToolCallItem.
  suppress?(item: ItemModel): boolean;
  // openBesidePath returns the single file path this tool card references (its
  // file arg), or undefined when it references none - the ONLY tools that opt in
  // are the single-file ones (read_file/edit_file/write_file, floor §3.7;
  // multi-target apply_patch and directory/pattern grep/ls/glob are excluded).
  // ToolCallItem turns a non-undefined path into an "open beside" affordance
  // (relativized against the session cwd, out-of-cwd gated - fileOpenBeside.tsx).
  openBesidePath?(item: ItemModel): string | undefined;
  // openBesideInline moves the "open beside" control from the END of the
  // summary to INLINE, right after the openBesidePath text inside it - the
  // one case today is read_file, whose summary quotes the path verbatim
  // ("Read <path> · lines N-M"), so the control lands between the file name
  // and the line range it opens. The path must literally appear inside
  // summary()'s own text (same "never a dead anchor" contract as
  // summaryLink); when it does not, ToolRow falls back to the end placement.
  openBesideInline?: boolean;
  // summarySuffix appends extra text to the collapsed row's summary, computed
  // from the FULL thread model rather than just this item - the one case
  // today is ask_user's "— answered: ..." recap (kata h70z), which lives in
  // a LATER, separate userMessage item this item alone can't see. Undefined
  // (every other tool) means no suffix. ToolCallItem re-derives this
  // reactively off a live model subscription, so a settled row's summary
  // updates the moment the answer arrives without item itself needing a new
  // identity.
  summarySuffix?(item: ItemModel, model: ThreadModel | undefined): string | undefined;
  // summaryHiddenWhenExpanded drops the row's summary line while the row is
  // open. The one case today is shell: its summary IS the raw one-line
  // command, and the expanded body already renders that same command
  // pretty-printed (ShellCommandBlock), so an open row would show the call
  // twice - the collapsed row keeps the summary, where it is the only glance
  // at the command. Undefined (every other tool) renders the summary in both
  // states, as before.
  summaryHiddenWhenExpanded?: boolean;
  // summaryLink, if present, is a URL that appears verbatim inside this
  // row's own summary() text and should render as a real, clickable link
  // rather than plain text - kata xw3t, the collapsed-row counterpart to
  // tcp9's expanded-body link (web_fetch's "Fetched <url> · N bytes"
  // collapsed-row summary was the one surface tcp9 deliberately left inert;
  // see that field's own kata for why). A parallel field rather than
  // widening summary()'s own return type to ReactNode: summary is ALSO
  // consumed as a plain string by summarySuffix's own concatenation above
  // and by ToolCallCluster's "N steps · ..." template, and ToolRow's
  // collapsed-state truncation (middleSplit) operates on summary as raw
  // characters, not markup - widening the whole contract would touch every
  // one of those for a link only one descriptor has today. Undefined (every
  // descriptor but web_fetch) renders the row exactly as before. When the
  // returned URL is not literally found inside summary(item)'s own text,
  // ToolRow renders the plain text unchanged - never a link pointing
  // somewhere the visible words don't say.
  summaryLink?(item: ItemModel): string | undefined;
}

// A tool call is failed when the wire carries an error/status failure or when
// the tool descriptor has a domain-specific failure signal. Shell's nonzero
// exit code is the important example: the command completed as a tool result,
// but the result itself still failed. Keeping this predicate beside the
// descriptor registry lets ToolCallItem and transcript grouping agree without
// duplicating the registry-specific rule.
export function toolCallFailed(item: ItemModel): boolean {
  const descriptor = toolRendererFor(item.toolName ?? "");
  return (
    (item.error !== undefined && item.error !== "") || item.status === "failed" || (descriptor.failed?.(item) ?? false)
  );
}

const registry: ToolRendererDescriptor[] = [];

export function registerToolRenderer(d: ToolRendererDescriptor): void {
  registry.push(d);
}

// DEFAULT_DESCRIPTOR mirrors legacy renderer-tools.js's own
// toolRenderers.__default__ / defaultRenderer (renderer-tools.js:773-778):
// used for any tool name with no registered descriptor.
const DEFAULT_DESCRIPTOR: ToolRendererDescriptor = {
  match: () => true,
  summary: (item) => item.toolName ?? "tool",
  // The generic tool glyph: MCP tools (and any other unregistered name) have
  // no family descriptor, so they all wear the wrench.
  icon: "wrench",
  body: (props) =>
    createElement(Fragment, null, createElement(MCPToolArguments, props), createElement(RawToolOutput, props)),
};

// toolRendererFor mirrors legacy renderer-tools.js's own toolRendererFor
// name (renderer-tools.js:17-19). Exact-string matches are checked across
// the WHOLE registry before any predicate, so a specific descriptor
// (e.g. "job_stop") always wins over a broader family predicate
// (e.g. name.startsWith("job_")) regardless of registration order; among
// same-kind matches (two predicates, or two identical exact strings), the
// first-registered one wins. Falls back to DEFAULT_DESCRIPTOR, never throws
// - an unregistered tool name is an everyday case (most tools have no
// dedicated descriptor until T3 lands), not a bug.
export function toolRendererFor(toolName: string): ToolRendererDescriptor {
  const exact = registry.find((d) => d.match === toolName);
  if (exact) return exact;
  const predicate = registry.find((d) => typeof d.match === "function" && d.match(toolName));
  return predicate ?? DEFAULT_DESCRIPTOR;
}
